analyzers/src/SonarAnalyzer.CSharp/Rules/DisposableTypesNeedFinalizers.cs

```diff
@@ -26,56 +26,50 @@
 using Microsoft.CodeAnalysis.Diagnostics;
 using SonarAnalyzer.Common;
 using SonarAnalyzer.Helpers;
+using StyleCop.Analyzers.Lightup;
 
 namespace SonarAnalyzer.Rules.CSharp
 {
     [DiagnosticAnalyzer(LanguageNames.CSharp)]
     [Rule(DiagnosticId)]
     public sealed class DisposableTypesNeedFinalizers : SonarDiagnosticAnalyzer
     {
-        internal const string DiagnosticId = "S4002";
+        private const string DiagnosticId = "S4002";
         private const string MessageFormat = "Implement a finalizer that calls your 'Dispose' method.";
 
-        private static readonly DiagnosticDescriptor rule
+        private static readonly DiagnosticDescriptor Rule
             = DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
 
-        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(rule);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(Rule);
 
-        private static readonly ImmutableArray<KnownType> nativeHandles =
-            ImmutableArray.Create(
-                KnownType.System_IntPtr,
-                KnownType.System_UIntPtr,
-                KnownType.System_Runtime_InteropServices_HandleRef
-            );
+        private static readonly ImmutableArray<KnownType> NativeHandles =
+            ImmutableArray.Create(KnownType.System_IntPtr,
+                                  KnownType.System_UIntPtr,
+                                  KnownType.System_Runtime_InteropServices_HandleRef);
 
-        protected override void Initialize(SonarAnalysisContext context)
-        {
+        protected override void Initialize(SonarAnalysisContext context) =>
             context.RegisterSyntaxNodeActionInNonGenerated(c =>
             {
-                var classDeclaration = (ClassDeclarationSyntax)c.Node;
-                var classSymbol = c.SemanticModel.GetDeclaredSymbol(classDeclaration);
+                var declaration = (TypeDeclarationSyntax)c.Node;
+                var classSymbol = c.SemanticModel.GetDeclaredSymbol(declaration);
 
-                if (classSymbol.Implements(KnownType.System_IDisposable) &&
-                    HasNativeHandleFields(classDeclaration, c.SemanticModel) &&
-                    !HasFinalizer(classDeclaration))
+                if (classSymbol.Implements(KnownType.System_IDisposable)
```

Comment:
_c.ContainingSymbol.Kind != SymbolKind.NamedType_ check is missing here

Reply:
Nice find. I've forgot about this.